lighthouse-cli/performance-experiment/server.js

```diff
@@ -84,28 +82,103 @@ function requestHandler(request, response) {
 }
 
 function reportRequestHandler(request, response) {
-  const html = reportGenerator.generateHTML(lhResults, 'cli');
+  const results = database.getResults(request.parsedUrl.query.id || 0);
+  if (!results) {
+    response.writeHead(404);
+    response.end('404: Resource Not Found');
+    return;
+  }
+  results.relatedReports = database.timeStamps.map((generatedTime, index) => {
+    return {reportUrl: `/?id=${index}`, url: database.url, generatedTime};
+  });
+
+  const reportGenerator = new ReportGenerator();
+  const html = reportGenerator.generateHTML(results, 'cli');
+
   response.writeHead(200, {'Content-Type': 'text/html'});
   response.end(html);
 }
 
 function rerunRequestHandler(request, response) {
+  const flags = database.getFlags(request.parsedUrl.query.id || 0);
+  if (!flags) {
+    response.writeHead(404);
+    response.end('404: Resource Not Found');
+    return;
+  }
+
   let message = '';
   request.on('data', data => message += data);
 
   request.on('end', () => {
     const additionalFlags = JSON.parse(message);
-    const flags = Object.assign({}, lhParams.flags, additionalFlags);
+    Object.assign(flags, additionalFlags);
 
-    lighthouse(lhParams.url, flags, lhParams.config).then(results => {
+    lighthouse(database.url, flags, database.config).then(results => {
       results.artifacts = undefined;
-      lhResults = results;
+      const id = database.saveData(flags, results);
       response.writeHead(200);
-      response.end();
+      response.end(`/?id=${id}`);
     });
   });
 }
 
+/**
+ *  Store all the experiment data (flags and results) on hard disk.
+ *  Call this.clear() to delete all the temp files.
+ */
+class ExperimentDatabase {
+  constructor(url, config) {
+    this._url = url;
+    this._config = config;
+
+    this._root = fs.mkdtempSync(`${__dirname}/experiment-data-`);
```

Comment:
Just to make it looks nicer. Since `fs.mkdtempSync` will adds some random characters as postfix.